server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/IndexBalanceAllocationDeciderTests.java

```diff
@@ -0,0 +1,326 @@
+ */
+
+package org.elasticsearch.cluster.routing.allocation.decider;
+
+import org.elasticsearch.cluster.ClusterInfo;
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ESAllocationTestCase;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.ProjectId;
+import org.elasticsearch.cluster.metadata.ProjectMetadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.cluster.node.DiscoveryNodeUtils;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.GlobalRoutingTable;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.cluster.routing.IndexShardRoutingTable;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.RoutingNodesHelper;
+import org.elasticsearch.cluster.routing.RoutingTable;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.ShardRoutingState;
+import org.elasticsearch.cluster.routing.TestShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.IndexVersion;
+import org.elasticsearch.index.shard.IndexLongFieldRange;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.cluster.metadata.IndexMetadata.SETTING_CREATION_DATE;
+import static org.elasticsearch.cluster.routing.TestShardRouting.shardRoutingBuilder;
+import static org.elasticsearch.common.settings.ClusterSettings.createBuiltInClusterSettings;
+
+public class IndexBalanceAllocationDeciderTests extends ESAllocationTestCase {
+
+    private DiscoveryNode indexNodeOne;
+    private DiscoveryNode indexNodeTwo;
+    private DiscoveryNode searchNodeOne;
+    private DiscoveryNode searchNodeTwo;
+    private DiscoveryNode masterNode;
+    private DiscoveryNode machineLearningNode;
+
+    private RoutingNode routingIndexNodeOne;
+    private RoutingNode routingIndexNodeTwo;
+    private RoutingNode routingSearchNodeOne;
+    private RoutingNode routingSearchNodeTwo;
+    private RoutingNode routingMasterNode;
+    private RoutingNode routingMachineLearningNode;
+
+    private List<DiscoveryNode> allNodes;
+    private int numberOfPrimaryShards;
+    private int replicationFactor;
+    private ClusterState clusterState;
+    private IndexMetadata indexMetadata;
+    private Settings settings;
+    private RoutingAllocation routingAllocation;
+    private IndexBalanceAllocationDecider indexBalanceAllocationDecider;
+
+    private void setup(boolean exceedThreshold) {
```

Comment:
Continuing my other comment thread, I'm hoping you can remove this boolean flag. I don't think you need to assign shards to ML / master nodes to test canAllocate.